### PR TITLE
prosody: new variables for prosody

### PR DESCRIPTION
**Pull request by aaronkvanmeerten**
This adds parameters for several prosody configuration options
This also adds support for adding global modules and adding global configuration
Uses tpl to write /config/prosody.cfg.lua

### CODE REVIEW

File: prosody/rootfs/defaults/conf.d/jitsi-meet.cfg.lua [hunk 1]
```diff
@@ -1,9 +1,13 @@
 admins = { "{{ .Env.JICOFO_AUTH_USER }}@{{ .Env.XMPP_AUTH_DOMAIN }}" }
 plugin_paths = { "/prosody-plugins/", "/prosody-plugins-custom" }
 http_default_host = "{{ .Env.XMPP_DOMAIN }}"
-
+ 
 {{ $ENABLE_AUTH := .Env.ENABLE_AUTH | default "0" | toBool }}
 {{ $AUTH_TYPE := .Env.AUTH_TYPE | default "internal" }}
+{{ $JWT_ASAP_KEYSERVER := .Env.JWT_ASAP_KEYSERVER | default "" }}
+{{ $JWT_ALL_EMPTY := .Env.JWT_ALLOW_EMPTY | default "0" | toBool }}
+{{ $JWT_AUTH_TYPE := .Env.JWT_AUTH_TYPE | default "token" }}
```
Comment (saghul) [inline prosody/rootfs/defaults/conf.d/jitsi-meet.cfg.lua]: why is this needed?

File: prosody/rootfs/defaults/conf.d/jitsi-meet.cfg.lua [hunk 2]
```diff
@@ -1,9 +1,13 @@
 admins = { "{{ .Env.JICOFO_AUTH_USER }}@{{ .Env.XMPP_AUTH_DOMAIN }}" }
 plugin_paths = { "/prosody-plugins/", "/prosody-plugins-custom" }
 http_default_host = "{{ .Env.XMPP_DOMAIN }}"
-
+
```
Comment (saghul) [inline prosody/rootfs/defaults/conf.d/jitsi-meet.cfg.lua]: nit: extraneous space

File: prosody/rootfs/defaults/conf.d/jitsi-meet.cfg.lua [hunk 3]
```diff
@@ -77,8 +85,8 @@ Component "{{ .Env.XMPP_MUC_DOMAIN }}" "muc"
         {{ if .Env.XMPP_MUC_MODULES }}
         "{{ join "\";\n\"" (splitList "," .Env.XMPP_MUC_MODULES) }}";
         {{ end }}
-        {{ if eq $AUTH_TYPE "jwt" }}
-        "token_verification";
+        {{ if .Env.JWT_ENABLE_TOKEN_AUTH | default "0" | toBool }}
```
Comment (saghul) [inline prosody/rootfs/defaults/conf.d/jitsi-meet.cfg.lua]: this check is incorrect, `JWT_ENABLE_TOKEN_AUTH` no longer exists

File: prosody/rootfs/defaults/conf.d/jitsi-meet.cfg.lua [hunk 4]
```diff
@@ -1,9 +1,13 @@
...
 plugin_paths = { "/prosody-plugins/", "/prosody-plugins-custom" }
 http_default_host = "{{ .Env.XMPP_DOMAIN }}"
-
+ 
 {{ $ENABLE_AUTH := .Env.ENABLE_AUTH | default "0" | toBool }}
 {{ $AUTH_TYPE := .Env.AUTH_TYPE | default "internal" }}
+{{ $JWT_ASAP_KEYSERVER := .Env.JWT_ASAP_KEYSERVER | default "" }}
+{{ $JWT_ALL_EMPTY := .Env.JWT_ALLOW_EMPTY | default "0" | toBool }}
+{{ $JWT_AUTH_TYPE := .Env.JWT_AUTH_TYPE | default "token" }}
+{{ $JWT_TOKEN_AUTH_MODULE := .Env.JWT_TOKEN_AUTH_MODULE | default "token_verification" }}
```
Comment (saghul) [inline prosody/rootfs/defaults/conf.d/jitsi-meet.cfg.lua]: please document this in the README and pass it in docker-compose too

File: prosody/rootfs/defaults/prosody.cfg.lua [hunk 5]
```diff
@@ -1,3 +1,5 @@
+{{ $LOG_LEVEL := .Env.LOG_LEVEL | default "info" }}
```
Comment (saghul) [inline prosody/rootfs/defaults/prosody.cfg.lua]: please document this and pass it in docker-compose

File: prosody/rootfs/defaults/prosody.cfg.lua [hunk 6]
```diff
@@ -70,6 +72,9 @@ modules_enabled = {
 		--"watchregistrations"; -- Alert admins of registrations
 		--"motd"; -- Send a message to users when they log in
 		--"legacyauth"; -- Legacy authentication. Only used by some old clients and bots.
+        {{ if .Env.GLOBAL_MODULES }}
```
Comment (saghul) [inline prosody/rootfs/defaults/prosody.cfg.lua]: please document this and pass it in docker-compose

File: prosody/rootfs/defaults/prosody.cfg.lua [hunk 7]
```diff
@@ -143,9 +148,13 @@ authentication = "internal_plain"
 --  Logs info and higher to /var/log
 --  Logs errors to syslog also
 log = {
-	{ levels = {min = "info"}, to = "console"};
+	{ levels = {min = "{{ $LOG_LEVEL }}"}, to = "console"};
 }
 
+{{ if .Env.GLOBAL_CONFIG }}
```
Comment (saghul) [inline prosody/rootfs/defaults/prosody.cfg.lua]: please document this and pass it in docker-compose

File: prosody/rootfs/etc/cont-init.d/10-config [hunk 8]
```diff
@@ -31,6 +31,7 @@ fi
 
 if [[ ! -f $PROSODY_CFG ]]; then
     cp -r /defaults/* /config
+    tpl /defaults/prosody.cfg.lua > $PROSODY_CFG
```
Comment (saghul) [inline prosody/rootfs/etc/cont-init.d/10-config]: at this point you can remove the `cp -r` since we are templating all files. a `mkdir` would be necessary though.

File: prosody/rootfs/defaults/conf.d/jitsi-meet.cfg.lua (hunk 1 above)
Comment (aaronkvanmeerten) [reply prosody/rootfs/defaults/conf.d/jitsi-meet.cfg.lua]: The ASAP_KEYSERVER is used during ASAP authorization, where the public key is fetched from a keyserver.  The ALLOW_EMPTY is used to allow anonymous users when they have no JWT, while still validating incoming JWTs when provided.  The AUTH_TYPE is to control which JWT authorization module to use for incoming JWTs.

File: prosody/rootfs/defaults/conf.d/jitsi-meet.cfg.lua (hunk 4 above)
Comment (aaronkvanmeerten) [reply prosody/rootfs/defaults/conf.d/jitsi-meet.cfg.lua]: done

File: prosody/rootfs/defaults/prosody.cfg.lua (hunk 5 above)
Comment (aaronkvanmeerten) [reply prosody/rootfs/defaults/prosody.cfg.lua]: done

File: prosody/rootfs/defaults/prosody.cfg.lua (hunk 7 above)
Comment (aaronkvanmeerten) [reply prosody/rootfs/defaults/prosody.cfg.lua]: done

File: prosody/rootfs/defaults/prosody.cfg.lua (hunk 6 above)
Comment (aaronkvanmeerten) [reply prosody/rootfs/defaults/prosody.cfg.lua]: done

File: prosody/rootfs/etc/cont-init.d/10-config (hunk 8 above)
Comment (aaronkvanmeerten) [reply prosody/rootfs/etc/cont-init.d/10-config]: I'd argue the cp -r is more generally useful as it can continue to handle new files without any further changes.  I agree it's not doing much currently though, so I'm open to changing it if it's agreed that's better.

File: prosody/rootfs/defaults/conf.d/jitsi-meet.cfg.lua (hunk 3 above)
Comment (aaronkvanmeerten) [reply prosody/rootfs/defaults/conf.d/jitsi-meet.cfg.lua]: resolved, good catch.

**Comment by saghul** [comment]
Thanks @aaronkvanmeerten ! Pedantic note for next time :-) Please rebase instead of merging dev into your working branch.
